kubespawner/spawner.py

```diff
@@ -2409,44 +2397,64 @@ def _start_reflector(
         """
-        key = kind
-        ReflectorClass = reflector_class
 
-        def on_reflector_failure():
-            self.log.critical(
-                "%s reflector failed, halting Hub.",
-                key.title(),
-            )
-            sys.exit(1)
+        if self.enable_user_namespaces:
+            # One reflector for all namespaces (requires ClusterRole).
+            # Stop JupyterHub if something went wrong
+
+            def on_reflector_failure():
+                self.log.critical(
+                    "%s reflector failed, halting Hub.",
+                    key.title(),
+                )
+                sys.exit(1)
+
+            async def catch_reflector_start(func):
+                try:
+                    await func
+                except Exception:
+                    self.log.exception(f"Reflector for {kind} failed to start.")
+                    sys.exit(1)
+
+        else:
+            # Created a dedicated reflector for each namespace (does not require ClusterRole).
+            # Namespace can be already deleted, or there can be no access a specific namespace, but not others,
+            # so halting the entire Jupyterhub instance is not the best idea.
+            # Instead, just halt the failing spawner.
+
+            on_reflector_failure = None
+
+            async def catch_reflector_start(func):
+                try:
+                    await func
+                except Exception:
+                    self.log.exception(f"Reflector for {kind} failed to start.")
+                    raise
 
-        previous_reflector = self.__class__.reflectors.get(key)
+        key = self._get_reflector_key(kind)
+        previous_reflector = self.__class__.reflectors.get(key, None)
 
         if replace or not previous_reflector:
-            self.__class__.reflectors[key] = ReflectorClass(
+            self.__class__.reflectors[key] = reflector_class(
                 parent=self,
                 namespace=self.namespace,
                 on_failure=on_reflector_failure,
                 **kwargs,
             )
-            f = asyncio.ensure_future(self.__class__.reflectors[key].start())
-
-            async def catch_reflector_start():
-                try:
-                    await f
-                except Exception:
-                    self.log.exception(f"Reflector for {kind} failed to start.")
-                    sys.exit(1)
-
-            asyncio.create_task(catch_reflector_start())
+            await catch_reflector_start(self.__class__.reflectors[key].start())
 
         if replace and previous_reflector:
             # we replaced the reflector, stop the old one
-            asyncio.ensure_future(previous_reflector.stop())
+            await previous_reflector.stop()
+
+        reflector = self.__class__.reflectors[key]
+        if not reflector.first_load_future.done():
+            await catch_reflector_start(reflector.first_load_future)
```

Comment:
I think this section isn't needed.

In reflector.py I see the start function `await _list_and_update()` in a `try` block. I think that either the `start()` function will have raised an error, or it will have set `first_load_future` already.

Reply:
Fixed